// dependency-lint/dist/formatters/summary_formatter.js
(function() {
  var SummaryFormatter, _, colors, errorMessages;

  _ = require('lodash');

  colors = require('colors/safe');

  errorMessages = require('./error_messages');

  SummaryFormatter = (function() {
    function SummaryFormatter(arg) {
      this.minimal = arg.minimal, this.stream = arg.stream;
    }

    SummaryFormatter.prototype.print = function(arg) {
      var fixed, fixes, i, len, module, modules, ref, results, type;
      fixes = (ref = arg.fixes) != null ? ref : {}, results = arg.results;
      for (type in results) {
        modules = results[type];
        if (this.minimal) {
          modules = _.filter(modules, function(arg1) {
            var error, errorIgnored;
            error = arg1.error, errorIgnored = arg1.errorIgnored;
            return error && !errorIgnored;
          });
        }
        if (modules.length === 0) {
          continue;
        }
        this.write(type + ":");
        for (i = 0, len = modules.length; i < len; i++) {
          module = modules[i];
          fixed = _.includes(fixes[type], module.name);
          this.write(this.moduleOutput(module, fixed), 1);
        }
        this.write('');
      }
      if (!(this.minimal && this.errorCount(results) === 0)) {
        return this.write(this.summaryOutput(results));
      }
    };

    SummaryFormatter.prototype.moduleOutput = function(arg, fixed) {
      var error, errorIgnored, files, header, message, name, scripts;
      error = arg.error, errorIgnored = arg.errorIgnored, files = arg.files, name = arg.name, scripts = arg.scripts;
      if (error) {
        message = errorMessages[error];
        if (errorIgnored) {
          return colors.yellow("- " + name + " (" + message + " - ignored)");
        } else {
          header = fixed ? colors.magenta("✖ " + name + " (" + message + " - fixed)") : colors.red("✖ " + name + " (" + message + ")");
          return header + colors.gray(this.errorSuffix({
            files: files,
            scripts: scripts
          }));
        }
      } else {
        return (colors.green('✓')) + " " + name;
      }
    };

    SummaryFormatter.prototype.indent = function(str, count) {
      var i, prefix, ref;
      prefix = '';
      if (count > 0) {
        for (i = 1, ref = count; 1 <= ref ? i <= ref : i >= ref; 1 <= ref ? i++ : i--) {
          prefix += '  ';
        }
      }
      return prefix + str;
    };

    SummaryFormatter.prototype.write = function(data, indent) {
      if (indent == null) {
        indent = 0;
      }
      data = data.split('\n').map((function(_this) {
        return function(str) {
          return _this.indent(str, indent);
        };
      })(this)).join('\n') + '\n';
      return this.stream.write(data, 'utf8');
    };

    SummaryFormatter.prototype.errorCount = function(results) {
      var count, error, errorIgnored, i, len, modules, ref, title;
      count = 0;
      for (title in results) {
        modules = results[title];
        for (i = 0, len = modules.length; i < len; i++) {
          ref = modules[i], error = ref.error, errorIgnored = ref.errorIgnored;
          if (error && !errorIgnored) {
            count += 1;
          }
        }
      }
      return count;
    };

    SummaryFormatter.prototype.errorSuffix = function(usage) {
      var i, item, len, list, suffix, type;
      suffix = '';
      for (type in usage) {
        list = usage[type];
        if (!(list && list.length > 0)) {
          continue;
        }
        suffix += '\n' + this.indent("used in " + type + ":", 2);
        for (i = 0, len = list.length; i < len; i++) {
          item = list[i];
          suffix += '\n' + this.indent(item, 3);
        }
      }
      return suffix;
    };

    SummaryFormatter.prototype.summaryOutput = function(results) {
      var errors, msg, prefix;
      errors = this.errorCount(results);
      prefix = colors.green('✓');
      if (errors > 0) {
        prefix = colors.red('✖');
      }
      msg = prefix + " " + errors + " error";
      if (errors !== 1) {
        msg += 's';
      }
      return msg;
    };

    return SummaryFormatter;

  })();

  module.exports = SummaryFormatter;

}).call(this);
